OrbitSsh/Channel.cpp

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitSsh/Channel.h"
+
+#include <libssh2.h>
+
+#include <memory>
+#include <optional>
+#include <outcome.hpp>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitSsh/ResultType.h"
+
+namespace OrbitSsh {
+
+Channel::Channel(Session* session_ptr)
+    : session_ptr_(session_ptr), raw_channel_ptr_(nullptr) {}
+
+Channel::~Channel() { libssh2_channel_free(raw_channel_ptr_); }
+
+ResultType Channel::Read(std::string* result) {
+  int rc;
+  bool read_at_least_once = false;
+  int buffer_size = sizeof(buffer_);
+  do {
+    rc = libssh2_channel_read(raw_channel_ptr_, buffer_, buffer_size);
+
+    if (rc == LIBSSH2_ERROR_EAGAIN) break;
+    if (rc < 0) {
+      ERROR("Unable to read channel, error code %d", rc);
+      return kError;
+    }
+
+    *result += std::string(buffer_, rc);
+    read_at_least_once = true;
+  } while (rc == buffer_size);
+
+  if (read_at_least_once) return kSuccess;
+  return kAgain;
+}
+
+ResultType Channel::Write(const std::string& text, int* written_length) {
+  int rc =
+      libssh2_channel_write(raw_channel_ptr_, text.data() + *written_length,
+                            text.length() - *written_length);
+  if (rc > 0) {
+    *written_length += rc;
+    return kSuccess;
+  }
+  if (rc == LIBSSH2_ERROR_EAGAIN) return kAgain;
+
+  ERROR("Unable to write to channel, error code %d", rc);
+  return kError;
+}
+
+ResultType Channel::WriteBlocking(const std::string& text) {
+  int written_length = 0;
+  do {
+    ResultType result = Write(text, &written_length);
+    if (result == kError) return result;
+  } while (written_length < text.length());
+
+  return kSuccess;
+}
+
+ResultType Channel::Exec(const std::string& command) {
+  int rc = libssh2_channel_exec(raw_channel_ptr_, command.c_str());
+  if (rc == 0) return kSuccess;
+
+  if (rc == LIBSSH2_ERROR_EAGAIN) return kAgain;
+
+  ERROR("Unable to perform channel exec, error code %d", rc);
+  return kError;
+}
+
+ResultType Channel::OpenChannelSession() {
+  raw_channel_ptr_ =
+      libssh2_channel_open_session(session_ptr_->GetRawSessionPtr());
+  if (raw_channel_ptr_ != nullptr) return kSuccess;
+
+  int last_errno = libssh2_session_last_errno(session_ptr_->GetRawSessionPtr());
+  if (last_errno == LIBSSH2_ERROR_EAGAIN) return kAgain;
+
+  ERROR("Unable to open channel, last errorno %d", last_errno);
+  return kError;
+}
```

Comment:
I now did factories for channel, but it is currently not really nice, because of the non-blocking nature of channel initialization. 

A factory returns ResultType to indicate success/error/wait and the result of the factory (`std::unique_ptr`) is passed as a IN/OUT pointer which is modified in case of success.
This is especially ugly because of `std::unique<Channel>*`. 

This will be nicer when ResultType will be refactored to hold a value. (different PR)